File: Go/Linked List/141_Linked List Cycle.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func hasCycle(head *ListNode) bool {
	first, next := head, head
	for next != nil && next.Next != nil {
		first = first.Next
		next = next.Next.Next
		if first == next {
			return true
		}
	}
	return false
}

File: C++/Linked List/141_Linked_List_Cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
    bool hasCycle(ListNode *head)
    {
        ListNode *first = head, *next = head;
        while (next != NULL && next->next != NULL)
        {
            head = head->next;
            next = next->next->next;
            if (head == next)
                return true;
        }
        return false;
    }
};
